logdna/request.go

```diff
@@ -0,0 +1,139 @@
+package logdna
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type httpRequest func(string, string, io.Reader) (*http.Request, error)
+type bodyReader func(io.Reader) ([]byte, error)
+type jsonMarshal func(interface{}) ([]byte, error)
+type httpClientInterface interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// Client used to make HTTP requests to the configuration api
+type requestConfig struct {
+	serviceKey  string
+	httpClient  httpClientInterface
+	apiURL      string
+	method      string
+	body        interface{}
+	httpRequest httpRequest
+	bodyReader  bodyReader
+	jsonMarshal jsonMarshal
+}
+
+// AlertResponsePayload contains alert data returned from the config-api
+type AlertResponsePayload struct {
+	Channels []channelResponse `json:"channels,omitempty"`
+	Error    string            `json:"error,omitempty"`
+	Name     string            `json:"name,omitempty"`
+	PresetID string            `json:"presetid,omitempty"`
+}
+
+// newRequestConfig abstracts the struct creation to allow for mocking
+func newRequestConfig(pc *providerConfig, method string, uri string, body interface{}, mutators ...func(*requestConfig)) *requestConfig {
+	rc := &requestConfig{
+		serviceKey:  pc.serviceKey,
+		httpClient:  &http.Client{Timeout: 15 * time.Second},
+		apiURL:      fmt.Sprintf("%s%s", pc.Host, uri), // uri should have a preceding slash (/)
+		method:      method,
+		body:        body,
+		httpRequest: http.NewRequest,
+		bodyReader:  ioutil.ReadAll,
+		jsonMarshal: json.Marshal,
+	}
+
+	// Testing only; Allow mutations passed in by tests
+	for _, mutator := range mutators {
+		mutator(rc)
+	}
+	return rc
+}
+
+func (c *requestConfig) MakeRequest() ([]byte, error) {
+	payloadBuf := bytes.NewBuffer([]byte{})
+	if c.body != nil {
+		pbytes, err := c.jsonMarshal(c.body)
+		if err != nil {
+			return nil, err
+		}
+		payloadBuf = bytes.NewBuffer(pbytes)
+	}
+
+	req, err := c.httpRequest(c.method, c.apiURL, payloadBuf)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("servicekey", c.serviceKey)
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error during HTTP request: %s, %+v", err, c)
+	}
+	defer res.Body.Close()
+
+	body, err := c.bodyReader(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing HTTP response: %s, %+v", err, c)
+	}
+  if res.StatusCode != http.StatusOK {
```

Comment:
nit: indent

Reply:
hrm, `go fmt` should have taken care of that.  Will run it again.

Reply:
It did